streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java

```diff
@@ -583,16 +583,21 @@ private void runLoop() {
                 }
             }
 
-            // try to process one fetch record from each task via the topology, and also trigger punctuate
+            // try to process several records from each task via the topology, and also trigger punctuate
             // functions if necessary, which may result in more records going through the topology in this loop
             if (totalNumBuffered > 0 || polledRecords) {
                 totalNumBuffered = 0;
 
                 if (!activeTasks.isEmpty()) {
                     for (StreamTask task : activeTasks.values()) {
-
-                        totalNumBuffered += task.process();
-
+                        // process this task's records to completion before
+                        // context switching to another task
+                        while (true) {
```

Comment:
Do we need to now check if we should be committing in here now? Before i think we'd process one record for each task and then check if we need to commit. Now we will process however many records that have been buffered before we commit. Could be quite a lot of tasks, which could mean that we are not committing as frequently as desired. This could obviously lead to having lots of duplicates during failure scenarios.
 It probably doesn't make that much of a difference when the commit interval is large, but if it is small it might be significant.